Added e46 config and wiki

jawillis opened the pull request.

jawillis committed:
Added e46 config

jawillis committed:
Add e46 config and wiki page

jawillis committed:
Update e46.md

jawillis committed:
revert config.h

jawillis committed:
Merge branch 'master' of https://github.com/jawillis/RaceChronoDiyBleDevice

timurrrr commented on can_db/e46.md:
```diff
@@ -0,0 +1,42 @@
+The easiest places to do this are either behind the intrument cluster or at the steering angle sensor, 
+which is located at the base of the steering column near the firewall.
+
+A simple plug-in harness can be made by cutting both sides of this connector and a few inches of wiring 
```
It would be useful to add a photo which harness you're talking about, and which pins are CAN/ACC/GND.

[Not blocking the PR, just something to add later]

jawillis replied:
Photos added

timurrrr commented on can_db/e46.md:
```diff
@@ -0,0 +1,42 @@
+
+Channel name | CAN ID | Equation | Notes
+------------ | --- | -------- | -----
+Accelerator position (%) | 809 | `F * 100 / 255` |
```
Does this equation correctly calculate the sub-% values, or should you change to 100.0?
Alternatively, I'd just write this as a shorter `F / 2.55`.

timurrrr commented on can_db/e46.md:
```diff
@@ -0,0 +1,42 @@
+------------ | --- | -------- | -----
+Accelerator position (%) | 809 | `F * 100 / 255` |
+Brake position (%) | 809 | `G & 1 * 100` | Brake pressures are not available on the main CAN bus, so this is just an on/off signal from the brake switch.
+Steering angle | 501 | `if(b & 0x80, 0 - ((b & 0x7f)<<8) \| a, (b<<8) \| a) * 0.04394` | Positive value = turning right. You can swap the equation in the if statement if you prefer it the other way around.
```
Is this the same as `if(B & 0x80, -1, 1) * bitsToUIntLe(raw, 0, 15) * 0.04394` ?

Strange that they didn't use the two's complement for negative values.

timurrrr commented on configs/e46.h:
```diff
@@ -0,0 +1,76 @@
+uint8_t getUpdateRateDivider(uint32_t can_id) {
+  // This is sent over the CAN bus 140 times per second and carries:
+  // vehicle speed
+  // Let's go for ~10 per second
```
Why so low? It's one of the most important data channels.
I suggest at least 20 updates per second for the vehicle speed.

I use 25 on 86's.

jawillis replied:
I was having throughput issues in testing and settled on 10Hz.  I think this can be tuned further.

timurrrr commented on configs/e46.h:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  // This is sent over the CAN bus 140 times per second and carries:
+  // individual wheel speeds 
```
Ditto. Also, consider merging these two `if`s together.

timurrrr commented on configs/e46.h:
```diff
@@ -0,0 +1,76 @@
+
+  // This is sent over the CAN bus 100 times per second and carries:
+  // Steering angle 
+  // we want ~10.
```
I find it useful to have the steering angle updated more frequently, as it greatly simplifies synchronizing data to video.
But given that in this CAN bus most of the high-speed data channels seem to be on separate IDs, I can imagine that you run out of bandwidth sooner than I do on my 86...
I trust your judgement here, just wanted to bring up the data/video sync concern.

jawillis replied:
I didn't consider the sync issues.  Bandwidth is certainly a concern.  I'm hoping this can be improved with a different processor/radio (esp32?)

timurrrr commented on configs/e46.h:
```diff
@@ -0,0 +1,76 @@
+  // This is sent over the CAN bus 100 times per second and carries:
+  // engine RPM 
+  // we want ~10.
+  if (can_id == 0x1F0) {
```
You've meant 0x316 here?

jawillis replied:
Correct.  This should be 0x316.

timurrrr commented on configs/e46.h:
```diff
@@ -0,0 +1,76 @@
+  // This is sent over the CAN bus 140 times per second and carries:
+  // vehicle speed
+  // Let's go for ~10 per second
+  if (can_id == 0x153) {
```
(here and below) Consider using decimal values in the code (e.g. 339 instead of 0x153) to better match the table in the .md file.

Also, perhaps _I_ should update the `ft86_gen1.h` file as well... Notice how the `ft86_gen2.h` file is written differently.

timurrrr commented on configs/e46.h:
```diff
@@ -0,0 +1,76 @@
+  // Throttle Position
+  // we want ~10.
+  if (can_id == 0x329) {
+    return 10;
```
I suggest 25 or even 50 updates per second for the channels that have pedal data.

jawillis replied:
That would indeed be nice, bandwidth permitting.  I have considered consolidating some of this data from multiple messages into a single custom super-message if other platforms prove to be as bandwidth-limited as the nRF52

timurrrr commented on configs/e46.h:
```diff
@@ -0,0 +1,76 @@
+    return 10;
+  }
+
+  // This is sent over the CAN bus 100 times per second and carries:
```
That's a very fast refresh rate for the oil temperature! (nothing actionable here, just lol).

timurrrr commented on can_db/e46.md:
```diff
@@ -0,0 +1,42 @@
+However, the communication protocol in RaceChrono is smart enough to optimize
+Bluetooth usage if multiple channels share the same CAN ID. As a general rule,
+if a new channel has the same CAN ID as an existing channel (such as "Throttle
+position" using the same CAN ID 320 as "Accelerator position"), then adding it
```
CAN ID 320 for "Accelerator position" is an 86 thing, please update with something more relevant for E46.

jawillis replied:
Yes, a cut-and-paste casualty.  Will change.

timurrrr commented on can_db/e46.md:
```diff
@@ -0,0 +1,42 @@
+Brake position (%) | 809 | `G & 1 * 100` | Brake pressures are not available on the main CAN bus, so this is just an on/off signal from the brake switch.
+Steering angle | 501 | `if(b & 0x80, 0 - ((b & 0x7f)<<8) \| a, (b<<8) \| a) * 0.04394` | Positive value = turning right. You can swap the equation in the if statement if you prefer it the other way around.
+Speed | 339 | `bitsToUIntLe(raw, 11, 13) / 16 * 0.277778` | You may want to check the multiplier against an external GPS device, especially if running larger/smaller diameter tires
+Engine RPM | 790 | `((D<<8) \| C) / 6.4` |
```
`bytesToUIntLe(raw, 2, 2) / 6.4`?

timurrrr commented on can_db/e46.md:
```diff
@@ -0,0 +1,42 @@
+Air temperature | 1557 | `D` |
+Clutch position (%) | 809 | `D & 1 * 100` | on/off signal from the clutch switch
+Fuel level (l) | 1555 | `C & 0x7f` |
+Odometer | 1555 | `((B<<8) \| A) *10000` |
```
`bytesToUIntLe(raw, 0, 2) * 10000`?

timurrrr commented on configs/e46.h:
```diff
@@ -0,0 +1,76 @@
+// Customizations for BMW e46
```
Please add
`//#include "configs/e46.h"`
to `config.h` between lines 22 and 23.

timurrrr reviewed (changes requested):
Overall looks good, just suggesting some minor corrections and a couple of minor questions.

Thanks a lot for your work!

jawillis commented:
Thanks for your feedback!  I'll apply and test these changes next week when I'm back home.

jawillis committed:
Review comments

jawillis committed:
added e46.h to config.h

jawillis commented:
I've addressed your review comments, except for the ones related to update rates since I have not been able to get better rates with my existing nRF52-based setup.  I will revisit those values when I have the ESP32 version fully tested.

timurrrr commented on can_db/e46.md:
```diff
@@ -4,24 +4,30 @@ The CAN bus is not available at the OBD-II port, so it must be tapped into behin
 The easiest places to do this are either behind the intrument cluster or at the steering angle sensor, 
 which is located at the base of the steering column near the firewall.
 
+![Steering Angle Sensor Plug](../images/e46_sensor_plug.jpg)
```
Great photo, makes it clear where the harness is.

I just pushed 56a8228cba11c6db2ae00abbef45d314450e28d4 with a re-compressed version of that photo.
Sorry I grew up in dial-up time, and still trying to save a few kB when possible 😄
You can now remove the file from this PR.

timurrrr replied:
(please do keep the reference in the `.md` file though!)

timurrrr commented on can_db/e46.md:
```diff
@@ -4,24 +4,30 @@ The CAN bus is not available at the OBD-II port, so it must be tapped into behin
 Channel name | CAN ID | Equation | Notes
 ------------ | --- | -------- | -----
-Accelerator position (%) | 809 | `F * 100 / 255` |
+Accelerator position (%) | 809 | `F / 2.54` |
```
Umm, why `2.54` instead of `2.55`?
I presume the max value of `F` is `255`?

timurrrr commented on can_db/e46.md:
```diff
@@ -4,24 +4,30 @@ The CAN bus is not available at the OBD-II port, so it must be tapped into behin
+Accelerator position (%) | 809 | `F / 2.54` |
 Brake position (%) | 809 | `G & 1 * 100` | Brake pressures are not available on the main CAN bus, so this is just an on/off signal from the brake switch.
-Steering angle | 501 | `if(b & 0x80, 0 - ((b & 0x7f)<<8) \| a, (b<<8) \| a) * 0.04394` | Positive value = turning right. You can swap the equation in the if statement if you prefer it the other way around.
+Steering angle | 501 | `if(B & 0x80, -1, 1) * bitsToUIntLe(raw, 0, 15) * 0.04394` | Positive value = turning right. You can swap the equation in the if statement if you prefer it the other way around.
```
FYI I ended up flipping positive/negative in the BRZ/86 equations to match how RaceChrono visualizes the steering angle. You may also want to do left=positive.

timurrrr commented on configs/e46.h:
```diff
@@ -2,71 +2,71 @@
-  if (can_id == 0x153) {
-    return 14;
+  if (can_id == 339) {
+    return 7;
```
Need to update the comment above to say 20 per second?

timurrrr commented on configs/e46.h:
```diff
@@ -2,71 +2,71 @@
-  if (can_id == 0x1F0) {
-    return 14;
+  if (can_id == 496) {
+    return 7;
```
ditto

timurrrr commented on configs/e46.h:
```diff
@@ -2,71 +2,71 @@
-  if (can_id == 0x1F0) {
-    return 10;
+  if (can_id == 790) {
+    return 5;
```
ditto

timurrrr commented on can_db/e46.md:
```diff
@@ -4,24 +4,30 @@ The CAN bus is not available at the OBD-II port, so it must be tapped into behin
+Constant/switched power and ground can also be found at this connector.  The harness pictured does not include the 
+power/ground.  
+
+![CAN bus harness](../images/e46_harness.jpg)
```
It's very hard to see the wire colors on that photos, presumably due to incandescent light source?
Can you try to take a photo outside in bright sunlight, or a similar light source?

Also I'd appreciate it if you can shrink/compress the file to be <300 kB.

timurrrr replied:
If you don't have GIMP/Photoshop to shrink/compress the photo please let me know and I can help with that.

timurrrr commented:
Overall looks good, just a couple of minor comments.
